src/routing_table/mod.rs

```diff
@@ -268,15 +271,26 @@ impl<T: Binary + Clone + Copy + Debug + Default + Hash + Xorable> RoutingTable<T
             }
             let _ = groups.insert(prefix, HashSet::new());
         }
-        RoutingTable {
+        let result = RoutingTable {
             our_name: our_name,
             min_group_size: min_group_size,
             our_group_prefix: our_group_prefix,
             groups: groups,
             needed: HashSet::new(),
+        };
+        if result.satisfies_invariant() {
+            Ok(result)
+        } else {
+            Err(Error::InvariantViolation)
         }
     }
 
+    fn satisfies_invariant(&self) -> bool {
+        self.groups
+            .keys()
+            .all(|&x| x == self.our_group_prefix || self.our_group_prefix.is_neighbour(&x))
```

Comment:
That is only half the invariant: that each routing table entry is a neighbour (or the group itself).

Conversely, we should also check whether each neighbour is present in the routing table. I think even that is locally verifiable: Every address that differs in exactly one bit from our own must be matched by some group key.